### ISSUE TITLE
Use NTS for Geometries - Higher Priority

### ISSUE THREAD

**Issue by plamen-i** [Mapsui/Mapsui#506]
@pauldendulk I know from documentation that this is on the roadmap but currently it is marked "Later".

Following recent trend in EF Core development (NTS support for spatial data types in [PostgreSQL/PostGIS](https://www.nuget.org/packages/Npgsql.EntityFrameworkCore.PostgreSQL.NetTopologySuite/), [SQL Server](https://dotnet.myget.org/feed/aspnetcore-dev/package/nuget/Microsoft.EntityFrameworkCore.SqlServer.NetTopologySuite) and [SQLite/SpatiaLite](https://github.com/bricelam/EFCore/commit/0a9dc2829e39e174c31721ae57d269ed2be2017a)) it clearly becomes important to switch to NTS geometries supplied by EF Core so maps can be rendered without any geometry type conversions. Can this get some higher priority given the fact that this is the last piece of full-stack online/offline .NET Core GIS application architecture (Server GDB <-> EF Core <-> WebAPI/WFS <-> Client Rendering (Mapsui) <-> Local GDB (SQLite/SpatiaLite) <-(sync)-> Server GDB) which needs some breaking changes to align geometry types with others?

**Comment by pauldendulk** [comment]
I am eager to go to NTS as well but I can only take small steps unfortunately. [This](http://mapsui.github.io/Mapsui/documentation/Mapsui-2.0.html) is the roadmap for 2.0 that I am working on. Not that much to go but it always takes some time to get it stable. Version 2.0 includes .NET Standard which is an important feature as well. Limiting the Viewport is also much requested.

After the 2.0 release I intend to start on the migration to NTS and going to an MIT license.

**Comment by groege** [comment]
Hey, just wanted to check in and ask if there has been any movement, it seems like 2.0 just needs some fixes and polish right? Does this mean we'll see progress on the change soon (which will as far as I understand also allow for the MIT license right?)

**Comment by groege** [comment]
Also just as information about spatialite (because I'm working on the spatialite build for Android see here: https://github.com/bricelam/mod_spatialite-NuGet/issues/6)

There are two dependencies in the spatialite project, that are LGPL as well... 
While spatialite itself can be used with either GPL, LGPL or MPL, the MPL would allow to use the project as part of the app without the open source restrictions, the LGPL dependencies, if I understand correctly basically make a spatialite binary LGPL.

**Comment by pauldendulk** [comment]
@groege After Mapsui moves to NTS there is more work to be done to go MIT. It will take some time, so don't wait for that.

Could you tell me a bit more about SpatiaLite? Where can I find the work you do on SpatiaLite? How does it relate to NTS, is it an alternative or an extension? How does it relate to sqlite-net-pcl?

**Comment by groege** [comment]
It can be compiled containing sqlite or as a sqlite plugin.
If you know postgis, its basically the same functions just in sqlite
Here is a list of functions:
https://www.gaia-gis.it/spatialite-3.0.0-BETA/spatialite-sql-3.0.0.html

Hope this answers your questions :)

**Comment by pauldendulk** [comment]
Is this an open source project? Where is it?

**Comment by groege** [comment]
https://www.gaia-gis.it

**Comment by FObermaier** [comment]
What is the state of this. Is there an interest in me taking this issue?

**Comment by radderz** [comment]
I think it is worth the change to go to NTS as that is the most supported framework with capabilities to do all the things spatial in .NET. Microsoft now uses this is in EFCore as the spatial library instead of the old native library that was only supported on windows. It provides a lot of tools out of the box, like importing from Shapefiles and Transformations.

**Comment by pauldendulk** [comment]
Alright, let's set this into motion. I have not started work yet. It will not be a simple port, the main work will be in making design decision, but I will try to spend some time on it. Will make a start tomorrow.

What I was thinking of is to add a new project that has a VectorLayer based on NTS. While the core Mapsui project will have no dependency on a geometry library, but will have it's own Point and BoundingBox.

Before we start I will have to decide on the current pull requests.

**Comment by radderz** [comment]
@pauldendulk I can give a hand with NTS as well as I get time, I have used the library in quite a few projects now and it is a very good library. 

NTS also has it's own indexes built in, i.e. RTree indexes that enable faster lookups within a space. This could improve the performance of the renderer when it comes down to figuring out which things to draw if there are many features on the map.

I think the idea of splitting out the vector parts out from the main Mapsui is a good idea as for those that are just putting pins on a map, NTS is a bit overkill. It's only when Polygons and LineStrings etc are to be used that I think NTS would be useful instead of reinventing the wheel on all the possibly object types and the calculations.

**Comment by kubaszostak** [comment]
I would like to add some things to consider before such a big change. NTS is an API for **modelling** and **manipulating** geometries. NTS is manipulating geometry so their coordinates changing. That's why they have to use **mutable** classes for geometries. That's why things like [LineSegment.Length](https://github.com/NetTopologySuite/NetTopologySuite/blob/c838c5061aa0f13d1fa65f828a868d790302ec06/src/NetTopologySuite/Geometries/LineString.cs#L247), [Polygon.Area](https://github.com/NetTopologySuite/NetTopologySuite/blob/c838c5061aa0f13d1fa65f828a868d790302ec06/src/NetTopologySuite/Geometries/Polygon.cs#L289) or even [Geometry.Envelope](https://github.com/NetTopologySuite/NetTopologySuite/blob/c838c5061aa0f13d1fa65f828a868d790302ec06/src/NetTopologySuite/Geometries/Geometry.cs#L594) are recalculated on every call. They look like properties but they act like methods. I would say rather expensive methods.

On the other hand Mapsui is a **visualisation** library. It doesn't manipulate geometries at all. So in Mapsui could be **immutable**.  In that scenario things like BoundingBox or Length would be calculated once during geometry initialization or using [lazy property initialization](https://killalldefects.com/2020/01/15/getting-lazy-in-c/) pattern. If performance is important i would also go for [structs over classes](https://adamsitnik.com/Value-Types-vs-Reference-Types/) for storing coordinates. In c# You can't get anything faster than collection of readonly struct stored in an array.

**Comment by FObermaier** [comment]
> NTS is an API for modelling and manipulating geometries. NTS is manipulating geometry so their coordinates changing.

This is not correctfor at least 2 reasons:
1. NTS also computes the relation between geometries (intersects, contains, covers) in a very sophisticated manner. The  `NetTopologySuite.Geometries.Prepared` namespace provides __very__ fast 1:m geometry predicate checks.
2. In NTS geometries are considered being value types. While it is possible to manipulate instantiated geometries, it is not encouraged to do so. Operations like `Buffer`, `Union`, `Intersection`, `Difference`, `SymDifference` don't operate on input geometries.   

`Geometry.Envelope` does not compute the envelope on every call, it just builds a new geometry based on a (possibly) pre-computed [`EnvelopeInternal`](https://github.com/NetTopologySuite/NetTopologySuite/blob/c838c5061aa0f13d1fa65f828a868d790302ec06/src/NetTopologySuite/Geometries/Geometry.cs#L609-L617).    

How often `Geometry.Area` and `Geometry.Length` are used during visualization is not clear to me.

**Comment by FObermaier** [comment]
@charlenni, I claim that using NTS geometries within MapsUI is not slower than using the currently used geometries taken from the `SharpMap` project.

**Comment by charlenni** [comment]
@FObermaier I think, the question is not, if we could get the same speed, but if we could get it faster. Ok not we, but for me.

The question is, in which direction MapsUI should go? Should it go into a Swiss army knife, which could display anything you want on any device you want, or should it go in the direction of a fast visualization library. For the first, NTS is the best choice. For the second one, there are perhaps better ways. I think, this should be considered first and then changes took place.

I tried to create a vector map with NTS in the background for primitives, and it was too slow. Could be, that I didn't used the correct types from NTS for that, but was, what I found.

**Comment by FObermaier** [comment]
> I tried to create a vector map with NTS in the background for primitives, and it was too slow. Could be, that I didn't used the correct types from NTS for that, but was, what I found.

You have a link?

**Comment by charlenni** [comment]
Sorry, I didn't find it. Perhaps I deleted it.

But you don't have to test it. Assume you have to convert all coordinates of any geometry you use each 16 ms to the used renderer format and then create a path to draw, then you could assume, that this costs much time. And this happens each time you draw the screen, again and again.

**Comment by FObermaier** [comment]
If you implement a `CoordinateSequence` (and `CoordinateSequenceFactory`) working with arrays of `SKPoint`s you don't have to do that.

**Comment by pauldendulk** [comment]
I am currently looking into a way in which Mapsui could support both Mapsui.Geometries and NTS. This is for two reasons:

1. The migration to NTS is a big change. It should be possible to introduce it in smaller steps with most of the functionality working for every ['Latest Build'](https://mapsui.com/documentation/nuget-of-latest-build.html?) package. They should be usable for early adopters, like they are now, so that we get feedback from them. 
2. I think it should be possible to render things that are not part of a geometry inheritance tree. For instance, right now we have a 'Raster' class which was derived from IGeometry, but it is not a real Geometry. It should be independent. Another such thing could perhaps be a VectorTile. This allows for performance optimization specifically for the VectorTile.

**Comment by pauldendulk** [comment]
I created a PR to prepare for NTS and wrote a [proposal](https://github.com/Mapsui/Mapsui/issues/1118#issuecomment-825585115) on how to proceed. Please give feedback everyone. Specifically those who would like to contribute like @FObermaier and @radderz.

**Comment by radderz** [comment]
@pauldendulk I have done an initial code review of the current PR which looks ok, I haven't had a chance to test it though. I'll review the proposal soon

**Comment by pauldendulk** [comment]
> If you implement a `CoordinateSequence` (and `CoordinateSequenceFactory`) working with arrays of `SKPoint`s you don't have to do that.

@FObermaier The port to NTS has just been merged. My objective was to get everything functioning with NTS. I did not pay any attention to performance, and the way I use NTS may not always be appropriate. Any help is appreciated. Perhaps you could skim through the PR, you may see obvious mistakes. 

https://github.com/Mapsui/Mapsui/pull/1454/files

Thing that I encountered while porting.
- It is not possible to add or remove coordinates. While editing a geometry this means you need to recreated the Geometry from scratch. Is this the right way to go?
- It is possible to modify the X and Y values in the Geometry.Coordinates. I use this while projecting, scaling and rotating. Is this the right way to do this?
- You need to call GeometryChanged after altering the coordinates. Not sure if I do this everywhere. What is the consequence if I don't?

**Comment by FObermaier** [comment]
Thanks @pauldendulk for the follow up.

First let me answer your questions.

> * It is not possible to add or remove coordinates. While editing a geometry this means you need to recreated the Geometry from scratch. Is this the right way to go?

You are right, NTS geometries are considered immutable, so you have to build a new geometry everytime you add a coordinate. You can implement your own `CoordinateSequence` and `CoordinateSequenceFactory` that is based on a list and use that just for editing geometries. That sequence would have to know the geometry it belongs to so you can call `Geometry.GeometryChanged()` when the list is changed.

> * It is possible to modify the X and Y values in the Geometry.Coordinates. I use this while projecting, scaling and rotating. Is this the right way to do this?

No, if you want to change the underlying coordinates create a class that implements `IEntireCoordinateSequenceFilter` and within the implementation of `Filter(CoordinateSequence)` perform the transformation. The property `Done` would always return `false` and the property `GeometryChanged` would return if the coordinates are now any different than before.
To use the filter you need to call `Geometry.Apply(filter);` (visitor pattern, gang of four)

> * You need to call GeometryChanged after altering the coordinates. Not sure if I do this everywhere. What is the consequence if I don't?

If you go with the filter class mentioned above you should be all set. If you don't call `GeometryChanged`, the extent of the geometry (`EnvelopeInternal`) is not flushed.

Generally I don't like the way new geometries are created within the solution, IMHO it is not the right choice to use the constructor and it is officially discouraged. If I were to manage this project, I'd follow this path:
1. Right at startup, configure an `NetTopologySuite.NtsGeometryServices` instance and assign it to `NtsGeometryServices.Instance` or any other static field that you have access to whenever you need to create geometries (see: https://mybinder.org/v2/gh/NetTopologySuite/jupyter/HEAD?labpath=GettingStarted.ipynb). The `NtsGeometryServices` lets you create factories to then create geometries.
2. I note that `IMap` interface as a CRS property. Somehow I'd map that CRS value to a SRID value that I'd then use to create a geometry factory for the map.
3. Whenever you need to create a factory that has a differen CRS/SRID you can do that.
4. Remove the extension methods that create geometries base on `IEnumerable<Coordinate>`, use the appropriate `GeometryFactory.Create...` method instead.

**Comment by pauldendulk** [comment]
@FObermaier Thanks a lot for this extensive review. I will create a new issue out of it and will perhaps split is up later in separate issues.

**Comment by pauldendulk** [comment]
I created a new issue for further improvements so will close this issue.
